python/instrumentation/openinference-instrumentation-llama-index/src/openinference/instrumentation/llama_index/_handler.py

```diff
@@ -0,0 +1,507 @@
+from openinference.semconv.trace import (
+    DocumentAttributes,
+    EmbeddingAttributes,
+    MessageAttributes,
+    OpenInferenceMimeTypeValues,
+    OpenInferenceSpanKindValues,
+    RerankerAttributes,
+    SpanAttributes,
+    ToolCallAttributes,
+)
+from opentelemetry import context as context_api
+from opentelemetry import trace as trace_api
+from opentelemetry.context import _SUPPRESS_INSTRUMENTATION_KEY
+from opentelemetry.util.types import AttributeValue
+from pydantic import BaseModel, Extra, PrivateAttr
+from typing_extensions import assert_never
+
+from llama_index.core import QueryBundle, Response
+from llama_index.core.base.agent.types import BaseAgent
+from llama_index.core.base.base_retriever import BaseRetriever
+from llama_index.core.base.embeddings.base import BaseEmbedding
+from llama_index.core.base.llms.base import BaseLLM
+from llama_index.core.base.response.schema import (
+    RESPONSE_TYPE,
+    AsyncStreamingResponse,
+    PydanticResponse,
+    StreamingResponse,
+)
+from llama_index.core.instrumentation.event_handlers import BaseEventHandler
+from llama_index.core.instrumentation.events import BaseEvent
+from llama_index.core.instrumentation.events.agent import (
+    AgentChatWithStepEndEvent,
+    AgentChatWithStepStartEvent,
+    AgentRunStepEndEvent,
+    AgentRunStepStartEvent,
+    AgentToolCallEvent,
+)
+from llama_index.core.instrumentation.events.chat_engine import (
+    StreamChatDeltaReceivedEvent,
+    StreamChatErrorEvent,
+)
+from llama_index.core.instrumentation.events.embedding import (
+    EmbeddingEndEvent,
+    EmbeddingStartEvent,
+)
+from llama_index.core.instrumentation.events.llm import (
+    LLMChatEndEvent,
+    LLMChatInProgressEvent,
+    LLMChatStartEvent,
+    LLMCompletionEndEvent,
+    LLMCompletionStartEvent,
+    LLMPredictEndEvent,
+    LLMPredictStartEvent,
+    LLMStructuredPredictEndEvent,
+    LLMStructuredPredictStartEvent,
+)
+from llama_index.core.instrumentation.events.query import (
+    QueryEndEvent,
+    QueryStartEvent,
+)
+from llama_index.core.instrumentation.events.rerank import (
+    ReRankEndEvent,
+    ReRankStartEvent,
+)
+from llama_index.core.instrumentation.events.retrieval import (
+    RetrievalEndEvent,
+    RetrievalStartEvent,
+)
+from llama_index.core.instrumentation.events.span import (
+    SpanDropEvent,
+)
+from llama_index.core.instrumentation.events.synthesis import (
+    GetResponseEndEvent,
+    GetResponseStartEvent,
+    SynthesizeEndEvent,
+    SynthesizeStartEvent,
+)
+from llama_index.core.instrumentation.span import BaseSpan
+from llama_index.core.instrumentation.span_handlers import BaseSpanHandler
+from llama_index.core.schema import QueryType
+
+logger = logging.getLogger(__name__)
+
+
+class _Span(
+    BaseSpan,
+    extra=Extra.allow,
+    keep_untouched=(singledispatchmethod, property),
+):
+    _otel_span: trace_api.Span = PrivateAttr()
+    _span_kind: Optional[str] = PrivateAttr()
```

Comment:
should this just fall back to 'chain' for now?

Reply:
It is falling back to chain when the span ends (and span_kind is still None).